File: cpp/subsets_ii.cc
/*
 * Filename: subsets_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 04 Aug 2015 07:18:39 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<int> > subsetsWithDup(vector<int>& nums)
		{
			vector<vector<int> > res;

			if (nums.empty())
			{
				return res;
			}

			vector<int> v;
			res.push_back(v);
			sort(nums.begin(), nums.end());
			subset_with_dup_aux(nums, -1, res, v);

			return res;
		}

	private:
		void subset_with_dup_aux(
				vector<int>& nums,
				int i,
				vector<vector<int> >& res,
				vector<int>& v)
		{
			for (vector<int>::size_type j = i+1; j < nums.size(); j++)
			{
				if (j-i-1 && nums[j] == nums[j-1])
				{
					continue;
				}

				v.push_back(nums[j]);
				res.push_back(v);
				subset_with_dup_aux(nums, j, res, v);
				v.pop_back();
			}
		}
};

File: java/SubsetsII.java
/*
 * Filename: SubsetsII.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 27 May 2015 02:08:18 PM EDT
 */

public class Solution {
	public List<List<Integer>> subsetsWithDup(int[] nums) {
		List<List<Integer>> res = new ArrayList<List<Integer>>();

		if (nums == null || nums.length == 0) {
			return res;
		}

		Arrays.sort(nums);
		for (int i = 0; i <= nums.length; i++) {
			List<List<Integer>> tmp = combine(nums, nums.length, i);
			res.addAll(tmp);
		}

		res.add(new ArrayList<Integer>());
		return res;
	}

	private List<List<Integer>> combine(int[] nums, int n, int k) {
		List<List<Integer>> res = new ArrayList<List<Integer>>();

		if (n <= 0 || k <= 0) {
			return res;
		}

		if (k == 1) {
			for (int i = 0; i < n; i++) {
				if (i > 0 && nums[i] == nums[i-1]) {
					continue;
				}

				List<Integer> comb = new ArrayList<Integer>();
				comb.add(nums[i]);
				res.add(comb);
			}

			return res;
		}

		if (k == n) {
			List<Integer> comb = new ArrayList<Integer>();
			
			for (int i = 0; i < n; i++) {
				comb.add(nums[i]);
			}

			res.add(comb);
			return res;
		}

		List<List<Integer>> list1 = combine(nums, n-1, k);
		List<List<Integer>> list2 = combine(nums, n-1, k-1);

		res.addAll(list1);
		for (List<Integer> comb: list2) {
			comb.add(nums[n-1]);
			if (!res.contains(comb)) {
				res.add(comb);
			}
		}

		return res;
	}
}
